### PR TITLE
sd-device: send udev database version

### LINKED ISSUE #23799: Wrong current_tags from device created with udev_monitor_receive_device
**systemd version the issue has been seen with**
 > 249.11-0ubuntu4

**Used distribution**
 > Ubuntu Kinetic Kudu (development branch)

**Linux kernel version used** (`uname -a`)

 > `Linux z370m-d3h 5.15.0-27-generic #28-Ubuntu SMP Thu Apr 14 04:55:28 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > ...

**Unexpected behaviour you saw**
 > ...

**Steps to reproduce the problem**
 > …

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
…
```

`udev_device_get_current_tags_list_entry` returns always same tags as `udev_device_get_tags_list_entry` if udev device is created with `udev_monitor_receive_device`. Works correctly if device is created with `udev_device_new_from_syspath`.

Problem seems to be that `udev_device` created from monitor has `sd_device` with `database_version == 0`.

Pull request https://github.com/systemd/systemd/pull/17622 added `device_database_supports_current_tags` that checks if `database_version` is `>=1`. If check fails, current_tags just returns regular tags.

Same problem if device is created with `udev_device_new_from_environment`.

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #23799.

### CODE REVIEW

**Comment by yuwata** [comment]
The current tag feature was introduced in v247.

File: src/libsystemd/sd-device/device-private.c [hunk 1]
```diff
@@ -536,59 +535,57 @@ static int device_update_properties_bufs(sd_device *device) {
...
         assert(device);
-        assert(strv);
 
         r = device_update_properties_bufs(device);
         if (r < 0)
                 return r;
 
-        *strv = device->properties_strv;
+        if (ret)
+                *ret = device->properties_strv;
```
Comment (keszybz) [inline src/libsystemd/sd-device/device-private.c:602]: Hmm, this change is not described in the commit message. If it's done, I'd prefer it to be split out to a separate commit. But I don't think it makes much sense: if `ret==NULL`, `device_get_properties_strv` doesn't make much sense.

File: src/libsystemd/sd-device/device-private.c [hunk 2]
```diff
@@ -536,59 +535,57 @@ static int device_update_properties_bufs(sd_device *device) {
...
                 return r;
 
-        *nulstr = device->properties_nulstr;
-        *len = device->properties_nulstr_len;
-
+        if (ret_nulstr)
+                *ret_nulstr = device->properties_nulstr;
+        if (ret_len)
+                *ret_len = device->properties_nulstr_len;
         return 0;
```
Comment (keszybz) [inline src/libsystemd/sd-device/device-private.c]: Same here: this change seems marginally useful, but the change in semantics should be a separate commit.

**Review by keszybz — COMMENTED** [review]
Looks reasonable, apart from some minor squabbles about the first commit.

**Comment by poettering** [comment]
lgtm

**Comment by yuwata** [comment]
@keszybz and @poettering Thank you for your review. The first commit was split into two. PTAL.

File: src/libsystemd/sd-device/device-private.c [hunk 3]
```diff
@@ -517,16 +521,23 @@ int device_new_from_nulstr(sd_device **ret, uint8_t *nulstr, size_t len) {
...
 
         if (!device->properties_buf_outdated)
                 return 0;
 
+        /* append udev database version */
+        buf_nulstr = newdup(char, "UDEV_DATABASE_VERSION=1\0", STRLEN("UDEV_DATABASE_VERSION=1") + 2);
+        if (!buf_nulstr)
+                return -ENOMEM;
+
+        nulstr_len += STRLEN("UDEV_DATABASE_VERSION=1") + 1;
```
Comment (muktupavels) [inline src/libsystemd/sd-device/device-private.c]: Is there reason why version is hard-coded here? Would not it be better to use `LATEST_UDEV_DATABASE_VERSION` from `device-internal.h`?

Comment (yuwata) [reply src/libsystemd/sd-device/device-private.c]: Good catch. I will update. Thank you.

**Comment by keszybz** [comment]
LGTM.

**Comment by keszybz** [comment]
I'll give this a few days to bake in `main` before backporting.

**Pull request merged**